rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,742 @@
+get around this, the required fields from the header MUST be included in the message itself. A non-gogo clone does exist
+`teleport.header.v1.ResourceHeader`, however, to get the fields, embeded custom marshalling must be manually written.
+
+If a resource has associated secrets (password, private key, jwt, mfa device, etc.) they should be defined in a separate
+resource and stored in a separate key range in the backend. The traditional pattern of defining secrets inline and only
+returning them if a `with_sercrets` flag is provided cause a variety of problems and introduce opportunity for human
+error to accidentally include secrets when they should not have been. It would then be the responsibility of the caller
+to get both the base resource and the corresponding secret resource if required.
+
+There are many things to consider, and many ways to design a resource
+specification. To provide consistency and uniformity there are a few
+[standards](https://cloud.google.com/apis/design/standard_fields)
+and [design patterns](https://cloud.google.com/apis/design/design_patterns)
+that should be followed when possible. The most notable of the design patterns
+is
+[Bool vs. Enum vs. String](https://cloud.google.com/apis/design/design_patterns#bool_vs_enum_vs_string). There have been several occasions in the past where a particular field
+was not flexible enough which prevented behavior from being easily extended to
+support a new feature.
+
+### API
+
+All APIs should follow the following conventions that are largely based on
+the [Google API style guide](https://cloud.google.com/apis/design/standard_methods).
+
+#### Create
+
+The `Create` RPC takes a resource to be created and must also return the newly created resource so that any fields that
+are populated server side are provided to clients without requiring an additional call to `Get`.
+
+The request MUST fail and return a `trace.AlreadyExists` error if a matching resource is already present in the backend.
+
+```protobuf
+// Creates a new Foo resource in the backend.
+    rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);
+
+message CreateFooRequest {
+  // The desired Foo to be created.
+  Foo foo = 1;
+}
+
+message CreateFooResponse {
+  // The created Foo as it exists in the backend. The only differences between the returned
+  // resource and the requested resource are due to properties that are updated by the server.
+  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
+  Foo foo = 1;
+}
+```
+
+#### Update
+
+The `Update` RPC takes a resource to be updated and must also return the updated resource so that any fields that are
+populated server side are provided to clients without requiring an additional call to `Get`. If partial updates of a
+resource are desired, the request may contain
+a [FieldMask](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#fieldmask).
```

Comment:
Are field masks handled solely by the gRPC package or does the server needs to call some specific functions from the gRPC package to accommodate for field masks? If the latter, it'd be nice to have a link to an example server-side implementation.

Reply:
Ah, I see. I don't know why I thought that maybe gRPC provides a way to "automatically" transform field masks into actual field updates so that I don't have to write a function like `applyDeviceUpdateMask` – that wouldn't make much sense. 🤔

Reply:
It's a common question. If you dig you'll likely find some reflect trickery, but usually I just recommend people to write the for/switch. It's simple, "dumb" code.